### ISSUE TITLE
grad of logsumexp produces NaN

### ISSUE THREAD

**Issue by howsiyu** [jax-ml/jax#23842]
### Description

The following code
```python
import jax
import jax.numpy as jnp

jax.grad(jax.scipy.special.logsumexp)(jnp.array([0.0, 100.0]), where=jnp.array([True, False]))
```
outputs
```python
Array([ 1., nan], dtype=float32)
```

A workaround I'm currently using is
```python

def logsumexp(a, where):
    a_max = jnp.max(a, initial=-jnp.inf, where=where)
    return a_max + jnp.log(jnp.sum(jnp.exp(jnp.where(where, a - a_max, -jnp.inf))))

jax.grad(logsumexp)(jnp.array([0.0, 100.0]), where=jnp.array([True, False]))
```
which produces the desired output

```python
Array([1., 0.], dtype=float32)
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.33
jaxlib: 0.4.33
numpy:  2.1.0
python: 3.11.9 (main, Apr  6 2024, 17:59:24) [GCC 11.4.0]
```

**Comment by superbobry** [comment]
We had a fix for this in #22836, but had to roll it back due to internal breakages at Google.

**Comment by jakevdp** [comment]
It looks like this is a duplicate of #22398, so I'm going to close this. Thanks for raising the issue, hopefully we'll be able to re-land the fix.
